google/gax/utils/protobuf.py

```diff
@@ -0,0 +1,168 @@
+    if isinstance(pb_or_dict, Message):
+        answer = getattr(pb_or_dict, key, default)
+    elif isinstance(pb_or_dict, collections.Mapping):
+        answer = pb_or_dict.get(key, default)
+    else:
+        raise TypeError('Tried to fetch a key %s on an invalid object; '
+                        'expected a dict or protobuf message.')
+
+    # If the object we got back is our sentinel, raise KeyError; this is
+    # a "not found" case.
+    if answer is SENTINEL:
+        raise KeyError(repr(key))
+
+    # If a subkey exists, call this method recursively against the answer.
+    if subkey:
+        return get(answer, subkey)
+
+    # Return the value.
+    return answer
+
+
+def set(pb_or_dict, key, value):
+    """Set the given key on the object.
+
+    Args:
+        pb_or_dict: An object which may be either:
+            - A subclass of :class:`google.protobuf.message.Message`
+            - A subclass of :class:`collections.MutableMapping` (e.g. dict)
+        key (str): The key on the object in question.
+        value (Any): The value to set.
+
+    Raises:
+        TypeError: If pb_or_dict is not a Message or Mapping.
+    """
+    # Sanity check: Is our target object valid?
+    if not isinstance(pb_or_dict, (collections.MutableMapping, Message)):
+        raise TypeError('Tried to set a key %s on an invalid object; '
+                        'expected a dict or protobuf message.' % key)
+
+    # We may be setting a nested key. Resolve this.
+    key, subkey = _resolve_subkeys(key)
+
+    # If a subkey exists, then get that object and call this method
+    # recursively against it using the subkey.
+    if subkey is not None:
+        if isinstance(pb_or_dict, collections.MutableMapping):
+            pb_or_dict.setdefault(key, {})
+        return set(get(pb_or_dict, key), subkey, value)
+
+    # Attempt to set the value on the types of objects we know how to deal
+    # with.
+    if isinstance(pb_or_dict, collections.MutableMapping):
+        pb_or_dict[key] = value
+    else:  # isinstance(pb_or_dict, Message):
+        setattr(pb_or_dict, key, value)
+
+
+def setdefault(pb_or_dict, key, value):
+    """Set the key on the object to the value if the current value is falsy.
+
+    Because protobuf Messages do not distinguish between unset values and
+    falsy ones particularly well, this method treats any falsy value
+    (e.g. 0, empty list) as a target to be overwritten, on both Messages
+    and dictionaries.
+
+    Args:
+        pb_or_dict: An object which may be either:
+            - A subclass of :class:`google.protobuf.message.Message`
+            - A subclass of :class:`collections.MutableMapping` (e.g. dict)
+        key (str): The key on the object in question.
+        value (Any): The value to set.
+
+    Raises:
+        TypeError: If pb_or_dict is not a Message or Mapping.
+    """
+    if not get(pb_or_dict, key, default=None):
+        set(pb_or_dict, key, value)
+
+
+def _resolve_subkeys(key, separator='.'):
```

Comment:
You don't appear to configure `separator` anywhere, so why provide it as a kwarg?